File: sdk/python/strait/events.py
"""Strait Event Triggers SDK — Python client."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any
from urllib.parse import quote

import httpx


@dataclass
class WaitForEventOptions:
    """Options for waiting on an external event."""

    event_key: str
    timeout_secs: int = 3600
    notify_url: str | None = None


@dataclass
class SendEventOptions:
    """Options for sending an event to resolve a waiting trigger."""

    event_key: str
    payload: dict[str, Any] | None = None


@dataclass
class EventTrigger:
    """Represents a durable event trigger."""

    id: str
    event_key: str
    project_id: str
    source_type: str
    trigger_type: str
    status: str
    timeout_secs: int
    requested_at: str
    expires_at: str
    workflow_run_id: str | None = None
    workflow_step_run_id: str | None = None
    job_run_id: str | None = None
    request_payload: Any = None
    response_payload: Any = None
    received_at: str | None = None
    error: str | None = None
    notify_url: str | None = None
    notify_status: str | None = None
    sent_by: str | None = None


class EventsClient:
    """Event trigger client for the Strait SDK."""

    def __init__(self, base_url: str, headers: dict[str, str]) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = httpx.Client(headers=headers, timeout=30.0)

    def wait_for_event(self, run_id: str, options: WaitForEventOptions) -> EventTrigger:
        """Pause the current run and wait for an external event."""
        url = f"{self._base_url}/sdk/v1/runs/{run_id}/wait-for-event"
        body: dict[str, Any] = {
            "event_key": options.event_key,
            "timeout_secs": options.timeout_secs,
        }
        if options.notify_url:
            body["notify_url"] = options.notify_url

        resp = self._client.post(url, json=body)
        resp.raise_for_status()
        return EventTrigger(**resp.json())

    def send_event(self, options: SendEventOptions) -> EventTrigger:
        """Send an event to resolve a waiting trigger."""
        url = f"{self._base_url}/v1/events/{quote(options.event_key, safe='')}/send"
        body: dict[str, Any] = {}
        if options.payload:
            body["payload"] = options.payload

        resp = self._client.post(url, json=body)
        resp.raise_for_status()
        return EventTrigger(**resp.json())

    def get_event_trigger(self, event_key: str) -> EventTrigger:
        """Get an event trigger by its event key."""
        url = f"{self._base_url}/v1/events/{quote(event_key, safe='')}"
        resp = self._client.get(url)
        resp.raise_for_status()
        return EventTrigger(**resp.json())

    def cancel_event_trigger(self, event_key: str) -> EventTrigger:
        """Cancel a waiting event trigger."""
        url = f"{self._base_url}/v1/events/{quote(event_key, safe='')}"
        resp = self._client.delete(url)
        resp.raise_for_status()
        return EventTrigger(**resp.json())

    def close(self) -> None:
        """Close the underlying HTTP client."""
        self._client.close()

File: sdk/typescript/src/events.ts
import type { EventTrigger, SendEventOptions, WaitForEventOptions } from "./types";

/** Event trigger client for the Strait SDK. */
export class EventsClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(baseUrl: string, headers: Record<string, string>) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.headers = headers;
  }

  /**
   * Pause the current run and wait for an external event.
   * The run transitions to "waiting" status and is re-dispatched
   * when the event arrives with the event payload as checkpoint data.
   */
  async waitForEvent(runId: string, options: WaitForEventOptions): Promise<EventTrigger> {
    const url = `${this.baseUrl}/sdk/v1/runs/${runId}/wait-for-event`;
    const body = {
      event_key: options.eventKey,
      timeout_secs: options.timeoutSecs ?? 3600,
      notify_url: options.notifyUrl,
    };

    const resp = await fetch(url, {
      method: "POST",
      headers: { ...this.headers, "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const err = await resp.json().catch(() => ({ error: resp.statusText }));
      throw new Error(`waitForEvent failed (${resp.status}): ${err.error || resp.statusText}`);
    }

    return resp.json() as Promise<EventTrigger>;
  }

  /**
   * Send an event to resolve a waiting trigger.
   * Returns the updated trigger with status "received".
   */
  async sendEvent(options: SendEventOptions): Promise<EventTrigger> {
    const url = `${this.baseUrl}/v1/events/${encodeURIComponent(options.eventKey)}/send`;
    const body = options.payload ? { payload: options.payload } : {};

    const resp = await fetch(url, {
      method: "POST",
      headers: { ...this.headers, "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const err = await resp.json().catch(() => ({ error: resp.statusText }));
      throw new Error(`sendEvent failed (${resp.status}): ${err.error || resp.statusText}`);
    }

    return resp.json() as Promise<EventTrigger>;
  }

  /** Get an event trigger by its event key. */
  async getEventTrigger(eventKey: string): Promise<EventTrigger> {
    const url = `${this.baseUrl}/v1/events/${encodeURIComponent(eventKey)}`;

    const resp = await fetch(url, {
      method: "GET",
      headers: this.headers,
    });

    if (!resp.ok) {
      const err = await resp.json().catch(() => ({ error: resp.statusText }));
      throw new Error(`getEventTrigger failed (${resp.status}): ${err.error || resp.statusText}`);
    }

    return resp.json() as Promise<EventTrigger>;
  }

  /** Cancel a waiting event trigger. */
  async cancelEventTrigger(eventKey: string): Promise<EventTrigger> {
    const url = `${this.baseUrl}/v1/events/${encodeURIComponent(eventKey)}`;

    const resp = await fetch(url, {
      method: "DELETE",
      headers: this.headers,
    });

    if (!resp.ok) {
      const err = await resp.json().catch(() => ({ error: resp.statusText }));
      throw new Error(`cancelEventTrigger failed (${resp.status}): ${err.error || resp.statusText}`);
    }

    return resp.json() as Promise<EventTrigger>;
  }
}
